### DISCUSSION TITLE
Open Graph and Uno Web Assembly

### DISCUSSION THREAD

**Discussion by Wandtket** [unoplatform/uno#10130] [Q&A]
I'm trying to create an open graph preview for sharing to social media on Uno WASM.

Does anyone know the best way of going about this? I tried injecting it into the document header in the AppManifest.js but it doesn't appear to be doing it fast enough for the web scrapers.

Edit:
I believe I only need to add a build event that applies the correct meta tags to the index.html page.

[Here's the documentation I'm following](https://developers.facebook.com/docs/sharing/webmasters#markup)

**Comment by carldebilly** [comment]
@Wandtket You can use a package like https://github.com/ghorsey/OpenGraph-Net... but it may not work out-of-the-box for the following reasons:

1. You many need to manually download the content using HttpClient instead of the integrated downloader which is not tuned for WASM and can raise errors... something like:
    ``` csharp
    using var http = new HttpClient();
    var html = await http.GetStringAsync("https://www.imdb.com/title/tt0109830/");
    var parsedResult = OpenGraph.ParseHtml(html);
    var title = parsedResult.Title;
    ```
2. You may hit CORS restrictions (which is the case in the previous code). In this case, that means you CAN'T access those pages directly from code running in a browser, no matter if it's coded in WASM or JavaScript. To solve this, if you doesn't control the server (which would be the case here for Imdb), you'll need a kind of proxy somewhere, probably a server you control.

**Reply by Wandtket** [reply]
Sounds a bit roundabout,

From some [Documentation](https://developers.facebook.com/docs/sharing/webmasters#markup) I just needed to add meta tags to the head of the Uno platforms index.html.

Its just that that part gets auto generated in web assembly so I needed a way to add the tags to the build event.

**Reply by jeromelaban** [reply]
You can create your own `index.html` [using this configuration](https://github.com/unoplatform/Uno.Wasm.Bootstrap/blob/main/doc/features-additional-files.md).

Edit: I updated the link, the public docs are not yet up to date.

**Reply by Wandtket** [reply]
I got it to successfully use the index.html but I'm not sure how to make it load the rest of the application. 


Here's the csproj

```
<PropertyGroup>
		<OutputType>Exe</OutputType>
		<TargetFramework>net5.0</TargetFramework>
		<NoWarn>NU1701</NoWarn>
		<EnableDefaultContentItems>false</EnableDefaultContentItems>
		<WasmShellIndexHtmlPath>index.html</WasmShellIndexHtmlPath>
		<WasmPWAManifestFile>manifest.json</WasmPWAManifestFile>
	</PropertyGroup>
```


Here's the index.html

```
<!DOCTYPE html>
<html>
<head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="apple-mobile-web-app-status-bar-style" content="black-translucent" />

    <script type="text/javascript" src="./require.js"></script>
    <script type="module" src="./uno-bootstrap.js"></script>
    $(ADDITIONAL_CSS)
    $(ADDITIONAL_HEAD)
</head>
<body>
    <div id="uno-body" class="container-fluid uno-body">
        <div class="uno-loader"
             loading-position="bottom"
             loading-alert="none">

            <!-- Logo: change src to customize the logo -->
            <img class="logo"
                 src="Assets/SplashScreen.png"
                 title="Uno is loading your application" />

            <progress></progress>
            <span class="alert"></span>
        </div>
    </div>
    <noscript>
        <p>This application requires Javascript and WebAssembly to be enabled.</p>
    </noscript>
</body>
</html>
```

Any help is greatly appreciated, thank you.

**Reply by jeromelaban** [reply]
> but I'm not sure how to make it load the rest of the application.

What do you mean by that? Can you provide error messages, if any?

**Reply by Wandtket** [reply]
Console window is throwing me:

`uno-bootstrap.js:2 Uncaught ReferenceError: config is not defined at uno-bootstrap.js:2:1`

**Reply by jeromelaban** [reply]
Thanks. Indeed, the documentation is incorrect, I've [just updated it](https://github.com/unoplatform/Uno.Wasm.Bootstrap/blob/main/doc/features-additional-files.md) to allow for the version choice. You're likely going to have to use the 3.x version of that file.

**Reply by Wandtket** [reply]
@jeromelaban That did it! Thank you for the work you guys put into this, it really sped up my schedule.
